providers/implementations/digests/ml_dsa_mu_prov.c

```diff
@@ -0,0 +1,339 @@
+            dst->mdctx = EVP_MD_CTX_new();
+            if (dst->mdctx == NULL
+                    || !EVP_MD_CTX_copy_ex(dst->mdctx, src->mdctx))
+                goto err;
+        }
+        if (src->propq != NULL) {
+            dst->propq = OPENSSL_strdup(src->propq);
+            if (dst->propq == NULL)
+                goto err;
+        }
+    }
+    return dst;
+err:
+    mu_freectx(dst);
+    return NULL;
+}
+
+static int mu_init(void *vctx, const OSSL_PARAM params[])
+{
+    MU_CTX *ctx = (MU_CTX *)vctx;
+
+    if (ossl_unlikely(!ossl_prov_is_running()))
+        return 0;
+
+    if (ctx->mdctx != NULL)
+        EVP_MD_CTX_init(ctx->mdctx);
+    ctx->remaining = ctx->digest_len;
+    return mu_set_ctx_params(vctx, params);
+}
+
+static int mu_get_params(OSSL_PARAM params[])
+{
+    return ossl_digest_default_get_params(params, SHA3_BLOCKSIZE(256),
+                                          SHAKE256_SIZE, SHAKE_FLAGS);
+}
+
+static const OSSL_PARAM *mu_settable_ctx_params(ossl_unused void *ctx,
+                                                ossl_unused void *provctx)
+{
+    return ml_dsa_mu_set_ctx_params_list;
+}
+
+static int set_property_query(MU_CTX *ctx, const char *propq)
+{
+    OPENSSL_free(ctx->propq);
+    ctx->propq = NULL;
+    if (propq != NULL) {
+        ctx->propq = OPENSSL_strdup(propq);
+        if (ctx->propq == NULL)
+            return 0;
+    }
+    return 1;
+}
+
+static EVP_MD *shake_digest(MU_CTX *ctx)
+{
+    if (ctx->md == NULL)
+        ctx->md = EVP_MD_fetch(ctx->libctx, "SHAKE256", ctx->propq);
+    return ctx->md;
+}
+
+static int digest_public_key(MU_CTX *ctx, const uint8_t *pub, size_t publen)
+{
+    int ret;
+    EVP_MD *md;
+    EVP_MD_CTX *mdctx;
+
+    md = shake_digest(ctx);
+    if (md == NULL)
+        return 0;
+    mdctx = EVP_MD_CTX_new();
+    if (mdctx == NULL)
+        return 0;
+    ret = EVP_DigestInit_ex(mdctx, md, NULL)
+        && EVP_DigestUpdate(mdctx, pub, publen)
+        && EVP_DigestFinalXOF(mdctx, ctx->tr, sizeof(ctx->tr));
+    EVP_MD_CTX_free(mdctx);
+    return ret;
+}
+
+static int mu_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    MU_CTX *ctx = (MU_CTX *)vctx;
+    struct ml_dsa_mu_set_ctx_params_st p;
+
+    if (ctx == NULL || !ml_dsa_mu_set_ctx_params_decoder(params, &p))
+        return 0;
+
+    if (p.ctx != NULL) {
+        void *vp = ctx->context;
+
+        if (!OSSL_PARAM_get_octet_string(p.ctx, &vp, sizeof(ctx->context),
+                                         &(ctx->context_len))) {
+            ctx->context_len = 0;
+            return 0;
+        }
+    }
+    if (p.propq != NULL) {
+        if (p.propq->data_type != OSSL_PARAM_UTF8_STRING
+            || !set_property_query(ctx, p.propq->data))
+            return 0;
+    }
+    if (p.pubkey != NULL) {
+        if (p.pubkey->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+        if (!digest_public_key(ctx, p.pubkey->data, p.pubkey->data_size))
+            return 0;
+    }
```

Comment:
This is a required parameter, it looks to me (did I miss some relevant code) that if this is not set we'll go ahead anyway with 0-filled $`tr`$ value.